logger.go

```
@@ -27,6 +27,37 @@ import (
 	"time"
 )
 
+// A CheckedMessage is the result of a call to Logger.Check, which allows
+// especially performance-sensitive applications to avoid allocations for disabled
+// or heavily sampled log levels.
+type CheckedMessage struct {
+	logger Logger
+	used   uint32
+	lvl    Level
+	msg    string
+}
+
+// NewCheckedMessage constructs a CheckedMessage. It's only intended for use by
+// wrapper libraries, and shouldn't be necessary in application code.
+func NewCheckedMessage(logger Logger, lvl Level, msg string) *CheckedMessage {
+	return &CheckedMessage{
+		logger: logger,
+		lvl:    lvl,
+		msg:    msg,
+	}
+}
+
+// Write logs the pre-checked message with the supplied fields. It panics if
+// called more than once.
+func (m *CheckedMessage) Write(fields ...Field) {
+	if n := atomic.AddUint32(&m.used, 1); n > 1 {
+		// We could call logger.Log at a (possibly large) performance cost, but it's
+		// better to make this an obvious user error.
+		panic("Can't use a CheckedMessage more than once.")
```

only panic in `Development` -- in prod, we should silently ignore